be/src/vec/aggregate_functions/aggregate_function_window_funnel.h

```diff
@@ -0,0 +1,214 @@
+            int64_t timestamp = events[i].first;
+            int event_idx = events[i].second;
+            write_var_int(timestamp, out);
+            write_var_int(event_idx, out);
+        }
+    }
+
+    void read(BufferReadable& in) {
+        int64_t event_level;
+        read_var_int(event_level, in);
+        max_event_level = (int)event_level;
+        read_var_int(window, in);
+        int64_t size = 0;
+        read_var_int(size, in);
+        for (int64_t i = 0; i < size; i++) {
+            int64_t timestamp;
+            int64_t event_idx;
+
+            read_var_int(timestamp, in);
+            read_var_int(event_idx, in);
+            VecDateTimeValue time_value(timestamp);
+            add(time_value, (int)event_idx, max_event_level, window);
+        }
+    }
+};
+
+class AggregateFunctionWindowFunnel
+        : public IAggregateFunctionDataHelper<WindowFunnelState,
+                                              AggregateFunctionWindowFunnel> {
+public:
+    AggregateFunctionWindowFunnel(const DataTypes& argument_types_)
+            : IAggregateFunctionDataHelper<WindowFunnelState,
+                                           AggregateFunctionWindowFunnel>(argument_types_, {}) {
+    }
+
+    String get_name() const override { return "window_funnel"; }
+
+    DataTypePtr get_return_type() const override {
+        return std::make_shared<DataTypeInt32>();
+    }
+
+    void reset(AggregateDataPtr __restrict place) const override { this->data(place).reset(); }
+
+    void add(AggregateDataPtr __restrict place, const IColumn** columns, size_t row_num,
+             Arena*) const override {
+        const auto& window = static_cast<const ColumnVector<Int64>&>(*columns[0]).get_data()[row_num];
+        // handle mode in the future.
+        // be/src/olap/row_block2.cpp copy_data_to_column
+        const auto& timestamp = static_cast<const ColumnVector<VecDateTimeValue>&>(*columns[2]).get_data()[row_num];
+        const int NON_EVENT_NUM = 3;
+        for (int i = NON_EVENT_NUM; i < get_argument_types().size(); i++) {
+            const auto& is_set = static_cast<const ColumnVector<UInt8>&>(*columns[i]).get_data()[row_num];
+            if (is_set) {
+                this->data(place).add(timestamp, i - NON_EVENT_NUM,
+                                      get_argument_types().size() - NON_EVENT_NUM, window);
+            }
+        }
+    }
+
+    void merge(AggregateDataPtr __restrict place, ConstAggregateDataPtr rhs,
+               Arena*) const override {
+        this->data(place).merge(this->data(rhs));
+    }
+
+    void serialize(ConstAggregateDataPtr __restrict place, BufferWritable& buf) const override {
+        this->data(place).write(buf);
+    }
+
+    void deserialize(AggregateDataPtr __restrict place, BufferReadable& buf,
+                     Arena*) const override {
+        this->data(place).read(buf);
+    }
+
+    void insert_result_into(AggregateDataPtr __restrict place, IColumn& to) const override {
```

Comment:
Why change `ConstAggregateDataPtr` to `AggregateDataPtr`? Maybe we should use const_cast here to ensure `place` is a non-const pointer. What I concern is it will hurt performance for other agg functions.

Reply:
Why does it affect performance? const is just used in compile time?

Reply:
Not affect performance, it's my fault to express inaccurately. I mean it may affect other agg function's behavior and may incur some unexpected risk. So I think its better to keep its origin const type.